### PR TITLE
box-shadow is now serialized properly

### LINKED ISSUE #15203: Value of box-shadow is serialized in a wrong order
The order should be "horizontal offset, vertical offset, blur radius, spread distance", but in the current code, it is "blur radius, spread distance, horizontal offset, vertical offset". The parsing order is correct, which means if you input `1px 2px 3px 4px`, it is serialized to `3px 4px 1px 2px`...

Spec: https://drafts.csswg.org/css-backgrounds-3/#the-box-shadow

Relevant code: https://github.com/servo/servo/blob/master/components/style/properties/longhand/effects.mako.rs#L27-L44

### PR DESCRIPTION
**Pull request by niki-04**
Box shadow in now serialized properly. The changes are in the file components/style/properties/longhand/effects.mako.rs. Tests for the serialization of box-shadow are in tests/unit/style/properties/serialization.rs. ./mach test-unit -p style passes all tests including the newly added test case.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X ] These changes fix #15203 (github issue number if applicable).

- [X ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15703)

### CODE REVIEW

File: tests/unit/style/properties/serialization.rs [hunk 1]
```diff
@@ -1231,4 +1231,26 @@ mod shorthand_serialization {
...
         }
     }
+      mod effects {
+        pub use super::*;
+        pub use style::properties::longhands::box_shadow::SpecifiedValue as BoxShadow;
+        pub use style::values::specified::Shadow;
+        #[test]
+        fn box_shadow_should_serialize_correctly() {
+            let mut properties = Vec::new();
+            let color = Some(CSSColor {
```
Comment (emilio) [inline tests/unit/style/properties/serialization.rs]: Let's move the `property_declaration_block` function to the parent module, and write this test as follows, which will be easier to maintain:

```rust
let shadow_css = "box-shadow: 1px 2px 3px 4px";
let shadow = property_declaration_block(shadow_css);
assert_eq!(shadow.to_css_string(), shadow_css);
```

**Review by emilio — COMMENTED** [review]
Thanks for this PR! The serialization changes look great to me, though I believe the test can be shorter.

Also, could you squash your commits? We don't use to land commits like "Fixes tidy" or so.

With that test change, this can land with `r=emilio`.

Thanks again for this patch! :)

File: tests/unit/style/properties/serialization.rs (hunk 1 above)
Comment (niki-04) [inline tests/unit/style/properties/serialization.rs]: Thanks for the review. So how to use the property_declaration_block function how do I define it in the parent module?

Comment (emilio) [inline tests/unit/style/properties/serialization.rs]: You can just move it to the top of the file making it `pub` (above the first `mod`), and use it from the others.

File: tests/unit/style/properties/serialization.rs [hunk 2]
```diff
@@ -1231,4 +1231,26 @@ mod shorthand_serialization {
             assert_eq!(serialization, block_text);
         }
     }
+      mod effects {
```
Comment (emilio) [inline tests/unit/style/properties/serialization.rs]: Also, one nit, alignment seems to be off-by-one here.

Comment (niki-04) [inline tests/unit/style/properties/serialization.rs]: I did make the changes. Also committed and squashed the commits.

Comment (niki-04) [inline tests/unit/style/properties/serialization.rs]: I think something went wrong and it created merge master branch. I don't know how to undo that but last commit to be considered is in box-shadow is now serialized properly. Let me know if I need to do anything to fix it. Sorry if this messes something!!

**Comment by KiChjang** [comment]
Please remove the merge commit and rebase against master instead.

**Comment by niki-04** [comment]
I removed the merge commit. How do I rebase against the master instead.

**Comment by KiChjang** [comment]
First, `git fetch upstream master`, and then `git rebase upstream/master`. You will encounter merge conflicts along the way, I suggest using a mergetool like meld, and you can configure your git client to open it up to solve merge conflicts through `git mergetool`.

**Comment by Wafflespeanut** [comment]
@niki-04 Are you planning to finish this?

**Comment by niki-04** [comment]
@wafflespeanut sorry about the delay. I will finish it by today. Will let you know if i need help

**Comment by niki-04** [comment]
I was going through the changes I see property_declaration_block to parse_declaration_block. Should I make the relevant changes in my code.

**Comment by niki-04** [comment]
I resolved the conflicts and committed the code again. Let me know if I need to squash my commits. I did all the checks again everything passes :)

**Comment by Wafflespeanut** [comment]
No, you haven't? It looks like you haven't rebased it properly.

**Comment by niki-04** [comment]
When I rebased it showed me the conflicts and I resolved those and committed again.

**Comment by Wafflespeanut** [comment]
- Export the diff from your current branch. Since you need only your last 4 commits squashed, you can do something like `git diff HEAD~4.. > patch.diff`
 - `git checkout master`, sync with upstream and delete your old branch with `git branch -D box-shadow-serialization`
 - `git checkout -b box-shadow-serialization` and `git apply patch.diff` (resolve conflicts, if any, and also remove `patch.diff`)
 - Finally, `git add -A && git commit -m "<commit-message>"` and push it.

**Comment by niki-04** [comment]
@Wafflespeanut thanks for the information. I am trying that but when I apply my patch it fails with the error patch does not apply. I tried looking for the same error but couldn't find any relevant information.

**Comment by Wafflespeanut** [comment]
It may probably have conflicts. Check whether the files have conflicts. If they don't have any, then what does it say? Can you paste the log output?

**Comment by niki-04** [comment]
No conflicts. The output is
Checking patch components/style/properties/longhand/effects.mako.rs...
error: while searching for:
            if self.inset {
                try!(dest.write_str("inset "));
            }
            try!(self.blur_radius.to_css(dest));
            try!(dest.write_str(" "));
            try!(self.spread_radius.to_css(dest));
            try!(dest.write_str(" "));
            try!(self.offset_x.to_css(dest));
            try!(dest.write_str(" "));
            try!(self.offset_y.to_css(dest));

            if let Some(ref color) = self.color {
                try!(dest.write_str(" "));

error: patch failed: components/style/properties/longhand/effects.mako.rs:29
error: components/style/properties/longhand/effects.mako.rs: patch does not apply
Checking patch tests/unit/style/properties/serialization.rs...
error: while searching for:
            assert_eq!(serialization, block_text);
        }
    }

    mod keywords {
        pub use super::*;
        #[test]

error: patch failed: tests/unit/style/properties/serialization.rs:1130
error: tests/unit/style/properties/serialization.rs: patch does not apply

**Comment by Wafflespeanut** [comment]
Oh, okay. What if you try `git am -3 patch.diff`? From what I see in the manual, git should now try to merge the changes, and will result in the usual file conflicts.

**Comment by niki-04** [comment]
I get a Patch format detection failed.

**Comment by Wafflespeanut** [comment]
Oh wait, it should be done with `apply` (not `am`, which is only for exported patches using `format-patch`). Sorry about that. Try `git apply -3 --whitespace=fix`?

**Comment by niki-04** [comment]
git apply -3 --whitespace=fix doesn't work. It just hangs up and does nothing

**Comment by Wafflespeanut** [comment]
Oh, you need to point it to the patch like you did previously. `git apply -3 --whitespace=fix patch.diff`

**Comment by niki-04** [comment]
Oh my bad. It still doesn't work
 error: components/style/properties/longhand/effects.mako.rs: does not match index
error: tests/unit/style/properties/serialization.rs: does not match index

**Comment by Wafflespeanut** [comment]
Strange. I just pulled from your branch, exported the patch and applied it, and it seems to work. I'm unable to reproduce the error. Can you check the "Allow edits from maintainers" check box to the right? (so that I'll be able to push to your branch?)

**Comment by niki-04** [comment]
I did the check allow edits from maintainers. Thanks a lot for helping out. It is strange you don't get the error.

**Comment by Wafflespeanut** [comment]
What? I didn't close this!

**Comment by Wafflespeanut** [comment]
Gimme a moment. Something went wrong.

**Comment by Wafflespeanut** [comment]
@niki-04 Apologies. I'd accidentally pushed "nothingness" into your PR, and it's now closed. Anyway, I've opened this in #16104. Thanks for working on this :)

**Comment by niki-04** [comment]
No worries. I enjoyed working on this!!

**Pull request closed without merging**